src/rdkafka_msg.c

```diff
@@ -482,17 +482,18 @@ int rd_kafka_msgq_age_scan (rd_kafka_msgq_t *rkmq,
 			    rd_ts_t now) {
 	rd_kafka_msg_t *rkm, *tmp;
 	int cnt = rd_atomic32_get(&timedout->rkmq_msg_cnt);
+	int timedout_cnt = cnt;
 
 	/* Assume messages are added in time sequencial order */
 	TAILQ_FOREACH_SAFE(rkm, &rkmq->rkmq_msgs, rkm_link, tmp) {
 		if (likely(rkm->rkm_ts_timeout > now))
 			break;
 
 		rd_kafka_msgq_deq(rkmq, rkm, 1);
-		rd_kafka_msgq_enq(timedout, rkm);
+		timedout_cnt = rd_kafka_msgq_enq(timedout, rkm);
 	}
 
-	return rd_atomic32_get(&timedout->rkmq_msg_cnt) - cnt;
```

Comment:
Its O(N), but one would hope in practice nothing times out and its ~O(1).  aiui atomic operations are especially costly, so if we generally don't loop it seems like it may actually matter.  I'm pretty sure the calculation is already racey and shouldn't be used for something that must know exactly how many elements are in the queue at this exact moment, I think to know exactly you need to hold the queue lock.  The problem is that between when the atomic fetch is performed and the value is returned to the caller another thread could modify timedout.  Looking at the rdkafka_broker.c caller it seems ok with this raceyness, in the worst case you miss the new queue elements, but get them the next time through.  It looks like the only other caller only uses the return value in logging, so that's not terribly important.

That all said I'm happy to revert it doing the conservatively correct thing until we have numbers showing it matters.  Mostly this second patch was because I think you suggested it earlier.